File: 86-partition-list/ken/86.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """

        left_start = None
        left_end = None
        right_start = None
        right_end = None

        cur = head

        while cur != None:
            if cur.val < x:
                if left_start == None:
                    left_start = cur
                else:
                    left_end.next = cur

                left_end = cur
            else:
                if right_start == None:
                    right_start = cur
                else:
                    right_end.next = cur

                right_end = cur

            cur = cur.next

        if right_end != None:
            right_end.next = None

        if left_start != None:
            left_end.next = right_start

            return left_start

        return right_start

File: 86-partition-list/leochang/86.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* small = new ListNode(0);
        ListNode* big = new ListNode(0);
        ListNode* small_head = small;
        ListNode* big_head = big;
        while (head) {
            if (head->val < x) {
                small->next = head;
                small = small->next;
            } else {
                big->next = head;
                big = big->next;
            }
            head = head->next;
        }
        big->next = NULL;
        small->next = big_head->next;
        return small_head->next;
    }
};
